### PR TITLE
[REVIEW] Cache-machine

### PR DESCRIPTION
**Pull request by aocsa**
Issue #441 

- [x] BlazingTable:sizeInBytes
- [x] Unit Test Cache Machine 
- [x] DoubleSourceWaitingWorkerThread
- [x] SingleSourceWaitingWorkerThread
- [x] Graph Processor
- [x] E2E tests for project, join, filter and table_scan
- [x] E2E tests for order_by
- [x] E2E test with sort -> sort-sample-partition-merge 
- [x] parallel execution graph

### CODE REVIEW

File: engine/src/execution_graph/logic_controllers/LogicPrimitives.h [hunk 1]
```diff
@@ -83,5 +115,314 @@ TableViewPair createEmptyTableViewPair(std::vector<cudf::type_id> column_types,
...
+  std::queue<std::unique_ptr<ral::frame::BlazingTable> > gpuData;
+//  std::map<int,std::unique_ptr<CacheData> > operatingData;
+
+	std::vector<CacheDataType> cachePolicyTypes;
+  std::vector<unsigned long long> memoryPerCache;
+  std::vector<unsigned long long> usedMemory;
+  bool _finished;
+};
+
+static std::string type_string(cudf::data_type dtype) {
```
Comment (felipeblazing) [inline engine/src/execution_graph/logic_controllers/LogicPrimitives.h]: what is this used for?

File: engine/src/execution_graph/logic_controllers/LogicPrimitives.cpp [hunk 2]
```diff
@@ -73,5 +78,167 @@ TableViewPair createEmptyTableViewPair(std::vector<cudf::type_id> column_types,
 	return ret;
 }
 
-}
-}
+} // namespace frame
+
+namespace cache{
```
Comment (felipeblazing) [inline engine/src/execution_graph/logic_controllers/LogicPrimitives.cpp]: its very hard to tell where you made changes if you copy and paste stuff around like this. I am guessing only a few lines below this changed but it look slike all new code.

File: engine/src/execution_graph/logic_controllers/LogicPrimitives.h [hunk 3]
```diff
@@ -83,5 +115,314 @@ TableViewPair createEmptyTableViewPair(std::vector<cudf::type_id> column_types,
...
+        case TIMESTAMP_DAYS: return "TIMESTAMP_DAYS";
+        case TIMESTAMP_SECONDS: return "TIMESTAMP_SECONDS";
+        case TIMESTAMP_MILLISECONDS: return "TIMESTAMP_MILLISECONDS";
+        case TIMESTAMP_MICROSECONDS: return "TIMESTAMP_MICROSECONDS";
+        case TIMESTAMP_NANOSECONDS: return "TIMESTAMP_NANOSECONDS";
+        default: return "Unsupported type_id";
+    }
+}
 
+static void print_blazing_table_view(ral::frame::BlazingTableView table_view, const std::string table_name = ""){
```
Comment (felipeblazing) [inline engine/src/execution_graph/logic_controllers/LogicPrimitives.h]: why is this in here this seems like debugging/testing code

Comment (aocsa) [reply engine/src/execution_graph/logic_controllers/LogicPrimitives.h]: yes I removed this dcode

**Comment by kharoc** [comment]
*******************************************************************************
TOTAL SUMMARY for test suite: 
PASSED: 1219/1234
FAILED: 15/1234
CRASH: 0/1234
TOTAL: 1234
*******************************************************************************
!!!!!!!!!!!!!!!! FAILED gdf TESTS !!!!!!!!!!!!!!!!
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: TPCH Queries Test - gdf
    FAILED: 0/12 ['TEST_15']
    CRASH: 0/12
!!!!!!!!!!!!!!!! FAILED psv TESTS !!!!!!!!!!!!!!!!
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: TPCH Queries Test - psv
    FAILED: 0/12 ['TEST_15']
    CRASH: 0/12
!!!!!!!!!!!!!!!! FAILED parquet TESTS !!!!!!!!!!!!
!!!!!!!!!!!!!!!! FAILED orc TESTS !!!!!!!!!!!!!!!!
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Date Test - orc
    FAILED: 1/5 ['TEST_01']
    CRASH: 0/5
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Timestamp Test - orc
    FAILED: 1/12 ['TEST_02']
    CRASH: 0/12
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Bindable Alias Test - orc
    FAILED: 1/12 ['TEST_09']
    CRASH: 0/12
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Simple Distribution From Local Test - orc
    FAILED: 4/52 ['TEST_22', 'TEST_43', 'TEST_46', 'TEST_49']
    CRASH: 0/52
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: TPCH Queries Test - orc
    FAILED: 5/11 ['TEST_03', 'TEST_05', 'TEST_06', 'TEST_10', 'TEST_14', 'TEST_18']
    CRASH: 0/11
!!!!!!!!!!!!!!!! FAILED dask_cudf TESTS !!!!!!!!!!!!!!!!
MAX DELTA: 196.0
*******************************************************************************
Aggregations without group by Test:   Start Mem: 403.875   End Mem: 521.875   Diff: 118.0
Coalesce Test:   Start Mem: 521.875   End Mem: 447.875   Diff: -74.0
Column Basis Test:   Start Mem: 447.875   End Mem: 445.875   Diff: -2.0
Common Table Expressions Test:   Start Mem: 445.875   End Mem: 515.875   Diff: 70.0
Count without group by Test:   Start Mem: 515.875   End Mem: 539.875   Diff: 24.0
Date Test:   Start Mem: 539.875   End Mem: 505.875   Diff: -34.0
Timestamp Test:   Start Mem: 505.875   End Mem: 419.875   Diff: -86.0
Full outer join Test:   Start Mem: 419.875   End Mem: 419.875   Diff: 0.0
Group by Test:   Start Mem: 419.875   End Mem: 445.875   Diff: 26.0
Group by without aggregations Test:   Start Mem: 445.875   End Mem: 445.875   Diff: 0.0
Inner join Test:   Start Mem: 445.875   End Mem: 451.875   Diff: 6.0
Left outer join Test:   Start Mem: 451.875   End Mem: 451.875   Diff: 0.0
Non-EquiJoin Queries Test:   Start Mem: 451.875   End Mem: 545.875   Diff: 94.0
Nested Queries Test:   Start Mem: 545.875   End Mem: 499.875   Diff: -46.0
Order by Test:   Start Mem: 499.875   End Mem: 499.875   Diff: 0.0
Predicates With Nulls Test:   Start Mem: 499.875   End Mem: 499.875   Diff: 0.0
Simple String Test:   Start Mem: 499.875   End Mem: 443.875   Diff: -56.0
Unary ops Test:   Start Mem: 705.875   End Mem: 705.875   Diff: 0.0
Unify Tables Test:   Start Mem: 705.875   End Mem: 705.875   Diff: 0.0
Union Test:   Start Mem: 705.875   End Mem: 723.875   Diff: 18.0
Limit Test:   Start Mem: 723.875   End Mem: 453.875   Diff: -270.0
Where clause Test:   Start Mem: 453.875   End Mem: 453.875   Diff: 0.0
Bindable Alias Test:   Start Mem: 453.875   End Mem: 549.875   Diff: 96.0
Boolean Test:   Start Mem: 549.875   End Mem: 551.875   Diff: 2.0
Case Test:   Start Mem: 551.875   End Mem: 545.875   Diff: -6.0
Cast Test:   Start Mem: 545.875   End Mem: 451.875   Diff: -94.0
Concat Test:   Start Mem: 451.875   End Mem: 589.875   Diff: 138.0
Literal Test:   Start Mem: 589.875   End Mem: 579.875   Diff: -10.0
Dir Test:   Start Mem: 579.875   End Mem: 775.875   Diff: 196.0
File System Local Test:   Start Mem: 775.875   End Mem: 593.875   Diff: -182.0
File System S3 Test:   Start Mem: 593.875   End Mem: 645.875   Diff: 52.0
Like Test:   Start Mem: 645.875   End Mem: 755.875   Diff: 110.0
Substring Test:   Start Mem: 531.875   End Mem: 501.875   Diff: -30.0
Wild Card Test:   Start Mem: 501.875   End Mem: 573.875   Diff: 72.0
TPCH Queries Test:   Start Mem: 573.875   End Mem: 515.875   Diff: -58.0
Round Test:   Start Mem: 515.875   End Mem: 555.875   Diff: 40.0

**Comment by kharoc** [comment]
1 RAL

*******************************************************************************
TOTAL SUMMARY for test suite: 
PASSED: 1201/1234
FAILED: 33/1234
CRASH: 0/1234
TOTAL: 1234
*******************************************************************************
!!!!!!!!!!!!!!!! FAILED gdf TESTS !!!!!!!!!!!!!!!!
!!!!!!!!!!!!!!!! FAILED psv TESTS !!!!!!!!!!!!!!!!
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Full outer join Test - psv
    FAILED: 2/4 ['TEST_02', 'TEST_03']
    CRASH: 0/4
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Predicates With Nulls Test - psv
    FAILED: 2/5 ['TEST_02', 'TEST_03']
    CRASH: 0/5
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: File System Local Test - psv
    FAILED: 1/21 ['TEST_11']
    CRASH: 0/21
!!!!!!!!!!!!!!!! FAILED parquet TESTS !!!!!!!!!!!!
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Full outer join Test - parquet
    FAILED: 2/4 ['TEST_02', 'TEST_03']
    CRASH: 0/4
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Predicates With Nulls Test - parquet
    FAILED: 2/5 ['TEST_02', 'TEST_03']
    CRASH: 0/5
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: File System Local Test - parquet
    FAILED: 1/21 ['TEST_11']
    CRASH: 0/21
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Wild Card Test - parquet
    FAILED: 1/21 ['TEST_11']
    CRASH: 0/21
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: TPCH Queries Test - parquet
    FAILED: 0/12 ['TEST_15']
    CRASH: 0/12
!!!!!!!!!!!!!!!! FAILED orc TESTS !!!!!!!!!!!!!!!!
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Date Test - orc
    FAILED: 1/5 ['TEST_01']
    CRASH: 0/5
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Timestamp Test - orc
    FAILED: 1/12 ['TEST_02']
    CRASH: 0/12
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Full outer join Test - orc
    FAILED: 2/4 ['TEST_02', 'TEST_03']
    CRASH: 0/4
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Predicates With Nulls Test - orc
    FAILED: 3/5 ['TEST_02', 'TEST_03', 'TEST_04']
    CRASH: 0/5
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Bindable Alias Test - orc
    FAILED: 1/12 ['TEST_09']
    CRASH: 0/12
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: File System Local Test - orc
    FAILED: 1/21 ['TEST_11']
    CRASH: 0/21
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Simple Distribution From Local Test - orc
    FAILED: 4/52 ['TEST_22', 'TEST_43', 'TEST_46', 'TEST_49']
    CRASH: 0/52
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: Wild Card Test - orc
    FAILED: 1/21 ['TEST_11']
    CRASH: 0/21
    xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    SUMMARY for FAILED test suite: TPCH Queries Test - orc
    FAILED: 6/11 ['TEST_03', 'TEST_05', 'TEST_06', 'TEST_10', 'TEST_14', 'TEST_15', 'TEST_18']
    CRASH: 0/11
!!!!!!!!!!!!!!!! FAILED dask_cudf TESTS !!!!!!!!!!!!!!!!
MAX DELTA: 218.0
*******************************************************************************
Aggregations without group by Test:   Start Mem: 403.875   End Mem: 521.875   Diff: 118.0
Coalesce Test:   Start Mem: 521.875   End Mem: 507.875   Diff: -14.0
Column Basis Test:   Start Mem: 507.875   End Mem: 503.875   Diff: -4.0
Common Table Expressions Test:   Start Mem: 503.875   End Mem: 439.875   Diff: -64.0
Count without group by Test:   Start Mem: 439.875   End Mem: 473.875   Diff: 34.0
Date Test:   Start Mem: 473.875   End Mem: 515.875   Diff: 42.0
Timestamp Test:   Start Mem: 515.875   End Mem: 427.875   Diff: -88.0
Full outer join Test:   Start Mem: 427.875   End Mem: 427.875   Diff: 0.0
Group by Test:   Start Mem: 427.875   End Mem: 445.875   Diff: 18.0
Group by without aggregations Test:   Start Mem: 445.875   End Mem: 445.875   Diff: 0.0
Inner join Test:   Start Mem: 445.875   End Mem: 485.875   Diff: 40.0
Left outer join Test:   Start Mem: 485.875   End Mem: 485.875   Diff: 0.0
Non-EquiJoin Queries Test:   Start Mem: 485.875   End Mem: 471.875   Diff: -14.0
Nested Queries Test:   Start Mem: 471.875   End Mem: 435.875   Diff: -36.0
Order by Test:   Start Mem: 435.875   End Mem: 435.875   Diff: 0.0
Predicates With Nulls Test:   Start Mem: 435.875   End Mem: 555.875   Diff: 120.0
Simple String Test:   Start Mem: 555.875   End Mem: 501.875   Diff: -54.0
Unary ops Test:   Start Mem: 771.875   End Mem: 771.875   Diff: 0.0
Unify Tables Test:   Start Mem: 771.875   End Mem: 771.875   Diff: 0.0
Union Test:   Start Mem: 771.875   End Mem: 655.875   Diff: -116.0
Limit Test:   Start Mem: 655.875   End Mem: 475.875   Diff: -180.0
Where clause Test:   Start Mem: 475.875   End Mem: 475.875   Diff: 0.0
Bindable Alias Test:   Start Mem: 475.875   End Mem: 531.875   Diff: 56.0
Boolean Test:   Start Mem: 531.875   End Mem: 535.875   Diff: 4.0
Case Test:   Start Mem: 535.875   End Mem: 519.875   Diff: -16.0
Cast Test:   Start Mem: 519.875   End Mem: 573.875   Diff: 54.0
Concat Test:   Start Mem: 573.875   End Mem: 595.875   Diff: 22.0
Literal Test:   Start Mem: 595.875   End Mem: 585.875   Diff: -10.0
Dir Test:   Start Mem: 585.875   End Mem: 803.875   Diff: 218.0
File System Local Test:   Start Mem: 803.875   End Mem: 627.875   Diff: -176.0
File System S3 Test:   Start Mem: 627.875   End Mem: 551.875   Diff: -76.0
Like Test:   Start Mem: 551.875   End Mem: 765.875   Diff: 214.0
Substring Test:   Start Mem: 599.875   End Mem: 571.875   Diff: -28.0
Wild Card Test:   Start Mem: 571.875   End Mem: 525.875   Diff: -46.0
TPCH Queries Test:   Start Mem: 525.875   End Mem: 581.875   Diff: 56.0
Round Test:   Start Mem: 581.875   End Mem: 627.875   Diff: 46.0

**Comment by mario21ic** [comment]
rerun tests

**Comment by aocsa** [comment]
READY TO REVIEW!!

2 RAL

========================================================
DETAILED SUMMARY TESTS
========================================================
TestGroup                       InputType  Result 
Aggregations without group by   dask_cudf  Success     6
                                gdf        Success     6
                                orc        Success     6
                                parquet    Success     6
                                psv        Success     6
Bindable Alias                  dask_cudf  Success    12
                                gdf        Success    12
                                orc        Success    11
                                           Fail        1
                                parquet    Success    12
                                psv        Success    12
Boolean                         psv        Success    10
Case                            gdf        Success     9
                                           Fail        1
                                orc        Success    10
                                parquet    Success    10
                                psv        Success    10
Cast                            gdf        Success    10
                                orc        Success     9
                                parquet    Success    10
                                psv        Success    10
Coalesce                        dask_cudf  Success    11
                                gdf        Success    11
                                orc        Success    11
                                parquet    Success    11
                                psv        Success    11
Common Table Expressions        dask_cudf  Success     3
                                gdf        Success     3
                                orc        Success     3
                                parquet    Success     3
                                psv        Success     3
Concat                          dask_cudf  Success    10
                                gdf        Success    10
                                orc        Success    10
                                parquet    Success    10
                                psv        Success    10
Count without group by          dask_cudf  Success     6
                                gdf        Success     6
                                parquet    Success     6
                                psv        Success     6
Date                            dask_cudf  Success     5
                                gdf        Success     5
                                orc        Success     4
                                           Fail        1
                                parquet    Success     5
                                psv        Success     5
Dir                             gdf        Success    22
File System Local               dask_cudf  Success    21
                                orc        Success    21
                                parquet    Success    21
                                psv        Success    21
File System S3                  parquet    Success    21
                                psv        Success    21
Full outer join                 dask_cudf  Success     4
                                gdf        Success     4
                                orc        Success     4
                                parquet    Success     4
                                psv        Success     4
Group by                        dask_cudf  Success     5
                                gdf        Success     5
                                orc        Success     5
                                parquet    Success     5
                                psv        Success     5
Group by without aggregations   dask_cudf  Success     5
                                gdf        Success     5
                                orc        Success     5
                                parquet    Success     5
                                psv        Success     5
Inner join                      dask_cudf  Success     7
                                gdf        Success     7
                                orc        Success     7
                                parquet    Success     7
                                psv        Success     7
Left outer join                 dask_cudf  Success     4
                                gdf        Success     4
                                orc        Success     4
                                parquet    Success     4
                                psv        Success     4
Like                            gdf        Success     3
Limit                           dask_cudf  Success    10
                                gdf        Success    10
                                orc        Success    10
                                parquet    Success    10
                                psv        Success    10
Literal                         orc        Success     5
Nested Queries                  dask_cudf  Success     4
                                gdf        Success     4
                                orc        Success     4
                                parquet    Success     4
                                psv        Success     4
Non-EquiJoin Queries            dask_cudf  Success     8
                                gdf        Success     8
                                orc        Success     8
                                parquet    Success     8
                                psv        Success     8
Order by                        dask_cudf  Success     4
                                gdf        Success     4
                                orc        Success     4
                                parquet    Success     4
                                psv        Success     4
Predicates With Nulls           dask_cudf  Success     5
                                gdf        Success     5
                                orc        Success     5
                                parquet    Success     5
                                psv        Success     5
Round                           dask_cudf  Success     5
                                gdf        Success     5
                                orc        Success     5
                                parquet    Success     5
                                psv        Success     5
Simple Distribution From Local  dask_cudf  Success    52
                                gdf        Success    52
                                orc        Success    48
                                           Fail        4
                                parquet    Success    52
                                psv        Success    52
Simple String                   dask_cudf  Success     7
                                gdf        Success     7
                                orc        Success     7
                                parquet    Success     7
                                psv        Success     7
Substring                       dask_cudf  Success     8
                                gdf        Success     8
                                orc        Success     8
                                parquet    Success     8
                                psv        Success     8
TPCH Queries                    dask_cudf  Success    13
                                gdf        Success    13
                                orc        Fail        6
                                           Success     6
                                parquet    Success    13
                                psv        Success    12
                                           Fail        1
Tables from Pandas              gdf        Success    13
Timestamp                       orc        Success    11
                                           Fail        1
Unary ops                       dask_cudf  Success     6
                                gdf        Success     6
                                orc        Success     6
                                parquet    Success     6
                                psv        Success     6
Unify Tables                    dask_cudf  Success     9
                                gdf        Success     9
                                orc        Success     9
                                parquet    Success     9
                                psv        Success     9
Union                           dask_cudf  Success     2
                                gdf        Success     2
                                orc        Success     2
                                parquet    Success     2
                                psv        Success     2
Where clause                    dask_cudf  Success     9
                                gdf        Success     9
                                orc        Success     9
                                parquet    Success     9
                                psv        Success     9
Wild Card                       dask_cudf  Success    21
                                gdf        Success    21
                                orc        Success    21
                                parquet    Success    21
                                psv        Success    21
Name: Result, dtype: int64
========================================================
FAILED TESTS
========================================================
                        TestGroup InputType Result                                           TestId
0                  Bindable Alias       orc   Fail                                          TEST_09
1                            Case       gdf   Fail                                          TEST_13
2                            Date       orc   Fail                                          TEST_01
3  Simple Distribution From Local       orc   Fail                  TEST_22,TEST_43,TEST_46,TEST_49
4                    TPCH Queries       orc   Fail  TEST_03,TEST_05,TEST-08,TEST_10,TEST_14,TEST_18
5                    TPCH Queries       psv   Fail                                          TEST_15
6                       Timestamp       orc   Fail                                          TEST_02
*******************************************************************************
TOTAL SUMMARY for test suite: 
PASSED: 1453/1468
FAILED: 15/1468
CRASH: 0/1468
TOTAL: 1468
MAX DELTA: 537.0
*******************************************************************************
Aggregations without group by:   Start Mem: 832.875   End Mem: 1369.875   Diff: 537.0
Coalesce:   Start Mem: 1369.875   End Mem: 1341.875   Diff: -28.0
Common Table Expressions:   Start Mem: 1341.875   End Mem: 1377.875   Diff: 36.0
Count without group by:   Start Mem: 1377.875   End Mem: 1321.875   Diff: -56.0
Date:   Start Mem: 1321.875   End Mem: 1283.875   Diff: -38.0
Timestamp:   Start Mem: 1283.875   End Mem: 1379.875   Diff: 96.0
Full outer join:   Start Mem: 1379.875   End Mem: 1379.875   Diff: 0.0
Group by:   Start Mem: 1379.875   End Mem: 1389.875   Diff: 10.0
Group by without aggregations:   Start Mem: 1389.875   End Mem: 1385.875   Diff: -4.0
Inner join:   Start Mem: 1385.875   End Mem: 1361.875   Diff: -24.0
Left outer join:   Start Mem: 1361.875   End Mem: 1361.875   Diff: 0.0
Non-EquiJoin Queries:   Start Mem: 1361.875   End Mem: 1303.875   Diff: -58.0
Nested Queries:   Start Mem: 1303.875   End Mem: 1337.875   Diff: 34.0
Order by:   Start Mem: 1337.875   End Mem: 1337.875   Diff: 0.0
Predicates With Nulls:   Start Mem: 1337.875   End Mem: 1319.875   Diff: -18.0
Simple String:   Start Mem: 1319.875   End Mem: 1313.875   Diff: -6.0
Unary ops:   Start Mem: 1397.875   End Mem: 1451.875   Diff: 54.0
Unify Tables:   Start Mem: 1451.875   End Mem: 1407.875   Diff: -44.0
Union:   Start Mem: 1407.875   End Mem: 1383.875   Diff: -24.0
Limit:   Start Mem: 1383.875   End Mem: 1333.875   Diff: -50.0
Where clause:   Start Mem: 1333.875   End Mem: 1333.875   Diff: 0.0
Bindable Alias:   Start Mem: 1333.875   End Mem: 1393.875   Diff: 60.0
Boolean:   Start Mem: 1393.875   End Mem: 1409.875   Diff: 16.0
Case:   Start Mem: 1409.875   End Mem: 1387.875   Diff: -22.0
Cast:   Start Mem: 1387.875   End Mem: 1413.875   Diff: 26.0
Concat:   Start Mem: 1413.875   End Mem: 1475.875   Diff: 62.0
Literal:   Start Mem: 1475.875   End Mem: 1465.875   Diff: -10.0
Dir:   Start Mem: 1465.875   End Mem: 1555.875   Diff: 90.0
File System Local:   Start Mem: 1555.875   End Mem: 1519.875   Diff: -36.0
File System S3:   Start Mem: 1519.875   End Mem: 1459.875   Diff: -60.0
Like:   Start Mem: 1459.875   End Mem: 1567.875   Diff: 108.0
Substring:   Start Mem: 1513.875   End Mem: 1579.875   Diff: 66.0
Wild Card:   Start Mem: 1579.875   End Mem: 1601.875   Diff: 22.0
TPCH Queries:   Start Mem: 1601.875   End Mem: 1543.875   Diff: -58.0
Round:   Start Mem: 1543.875   End Mem: 1449.875   Diff: -94.0

File: comms/src/blazingdb/manager/Context.cc [hunk 4]
```diff
@@ -15,7 +15,15 @@ Context::Context(const uint32_t token,
       masterNode_{masterNode},
       logicalPlan_{logicalPlan},
       query_step{0},
-      query_substep{0} {}
+      query_substep{0},
+      kernel_id_{0} {}
+
+std::shared_ptr<Context> Context::clone() {
```
Comment (felipeblazing) [inline comms/src/blazingdb/manager/Context.cc]: Why is htere a clone here. Wouldn't properly implementing a copy constructor be the best way to handle this?

File: comms/src/blazingdb/manager/Context.cc [hunk 5]
```diff
@@ -56,14 +64,21 @@ std::string Context::getLogicalPlan() const { return logicalPlan_; }
 
 uint32_t Context::getContextToken() const { return token_; }
 
-uint32_t Context::getContextCommunicationToken() const { return query_substep; }
+std::string Context::getContextCommunicationToken() const {
+  return std::to_string(kernel_id_) + "_"  + std::to_string(query_substep) ;   
+}
 
 void Context::incrementQueryStep() {
+  std::unique_lock<std::mutex> lock(increment_step_mutex);
```
Comment (felipeblazing) [inline comms/src/blazingdb/manager/Context.cc]: how can two processes be in contention for incrementing the queryStep of a context? If that happens it feels like thats a bug right? Isn't this supposed to be linear?, if its not linear for some reason what does that mean?

Comment (felipeblazing) [reply comms/src/blazingdb/manager/Context.cc]: is the query subset not set back to 0 whe you move onto the next step?

Comment (felipeblazing) [reply comms/src/blazingdb/manager/Context.cc]: so we even need query steps?

**Pull request merged**